WMF Framework/WMFExploreFeedContentController.m

```diff
@@ -276,6 +286,220 @@ - (void)updateBackgroundSourcesWithCompletion:(void (^_Nonnull)(UIBackgroundFetc
+}
+
+- (WMFAppDefaultTabType)defaultTabType {
+    NSNumber *defaultTabTypeNumber = (NSNumber *)[self.exploreFeedPreferences objectForKey:WMFExploreFeedPreferencesDefaultTabTypeKey];
+    WMFAppDefaultTabType defaultTabType = (WMFAppDefaultTabType)[defaultTabTypeNumber intValue];
+    return defaultTabType;
+}
+
+- (BOOL)isDefaultTabExplore {
+    return (self.defaultTabType != WMFAppDefaultTabTypeExplore);
+}
+
+- (void)toggleContentGroupOfKind:(WMFContentGroupKind)contentGroupKind isOn:(BOOL)isOn {
+    NSSet *preferredSiteURLs = [NSSet setWithArray:self.preferredSiteURLs];
+    [self toggleContentGroupOfKind:contentGroupKind forSiteURLs:preferredSiteURLs isOn:isOn];
+}
+
+- (void)toggleContentGroupOfKind:(WMFContentGroupKind)contentGroupKind isOn:(BOOL)isOn forSiteURL:(NSURL *)siteURL {
+    [self toggleContentGroupOfKind:contentGroupKind forSiteURLs:[NSSet setWithObject:siteURL] isOn:isOn];
+}
+
+-(void)toggleContentForSiteURL:(NSURL *)siteURL isOn:(BOOL)isOn updateFeed:(BOOL)updateFeed {
+    [self updateExploreFeedPreferences:^(NSMutableDictionary *newPreferences, dispatch_block_t completion) {
+        NSString *key = siteURL.wmf_articleDatabaseKey;
+        if (isOn) {
+            [newPreferences setObject:[WMFExploreFeedContentController customizableContentGroupKindNumbers] forKey:key];
+        } else {
+            if ([newPreferences objectForKey:key]) {
+                [newPreferences removeObjectForKey:key];
+            }
+        }
+        completion();
+    } completion:^{
+        if (!updateFeed) {
+            return;
+        }
+        [self updateFeedSourcesUserInitiated:YES completion:nil];
+    }];
+}
+
+- (void)toggleContentGroupOfKind:(WMFContentGroupKind)contentGroupKind forSiteURLs:(NSSet<NSURL *> *)siteURLs isOn:(BOOL)isOn {
+    [self updateExploreFeedPreferences:^(NSMutableDictionary *newPreferences, dispatch_block_t completion) {
+        for (NSURL *siteURL in siteURLs) {
+            NSString *key = siteURL.wmf_articleDatabaseKey;
+            NSSet *oldVisibleContentSources = [newPreferences objectForKey:key];
+            NSMutableSet *newVisibleContentSources;
+
+            if (oldVisibleContentSources) {
+                newVisibleContentSources = [oldVisibleContentSources mutableCopy];
+            } else {
+                newVisibleContentSources = [NSMutableSet set];
+            }
+
+            if (isOn) {
+                [newVisibleContentSources addObject:@(contentGroupKind)];
+            } else {
+                [newVisibleContentSources removeObject:@(contentGroupKind)];
+            }
+
+            [newPreferences setObject:newVisibleContentSources forKey:key];
+            completion();
+        }
+    } completion:^{
+        [self updateFeedSourcesUserInitiated:YES completion:nil];
+    }];
+}
+
+- (void)updateExploreFeedPreferences:(void(^)(NSMutableDictionary *newPreferences, dispatch_block_t completion))update completion:(nullable dispatch_block_t)completion {
```

Comment:
Could this be `- (void)updateExploreFeedPreferences:(void(^)(NSMutableDictionary *newPreferences)) completion:(nullable dispatch_block_t)completion` - it seems like none of the blocks passed to this method are async and by requiring the method to be synchronous you can be certain you're still on the moc's private queue

Reply:
Do you mean lose the `dispatch_block_t completion` in `(NSMutableDictionary *newPreferences, dispatch_block_t completion))update`? 😬 I'm using that `completion` block to manage when `[moc wmf_setValue:newPreferences forKey:WMFExploreFeedPreferencesKey];` gets called, for example in here I am calling it while in the loop: https://github.com/wikimedia/wikipedia-ios/blob/aafc135e8d00c7d4dbbbaf07d80c9ba5c8d29150/WMF%20Framework/WMFExploreFeedContentController.m#L409 and here I'm calling it after the `if`: https://github.com/wikimedia/wikipedia-ios/blob/aafc135e8d00c7d4dbbbaf07d80c9ba5c8d29150/WMF%20Framework/WMFExploreFeedContentController.m#L380
Is there a better way to do this? 🤓

Reply:
For the one in the loop, you'll get the same result if you call it after the loop since it replaces the value each time it's called. At that point, you can remove the completion and just call `[moc wmf_setValue:newPreferences forKey:WMFExploreFeedPreferencesKey];` after `update(newPreferences);`

Reply:
Thanks 🙌🥂 also 🤦🏻‍♀️